crypto/cms/cms_sd.c

```diff
@@ -928,20 +917,25 @@ int CMS_SignerInfo_verify(CMS_SignerInfo *si)
         goto err;
 
     if (!cms_sd_asn1_ctrl(si, 1))
-        goto err;
+        return 0;
 
     alen = ASN1_item_i2d((ASN1_VALUE *)si->signedAttrs, &abuf,
                          ASN1_ITEM_rptr(CMS_Attributes_Verify));
     if (abuf == NULL || alen < 0)
         goto err;
-    r = EVP_DigestVerifyUpdate(mctx, abuf, alen);
-    OPENSSL_free(abuf);
-    if (r <= 0) {
-        r = -1;
-        goto err;
+    if (pknid == NID_ED25519 || pknid == NID_ED448) {
+        r = EVP_DigestVerify(mctx, si->signature->data, si->signature->length,
+                             abuf, alen) == 1;
```

Comment:
Drive by comment: Don't you need to free `abuf` in this case as per the other branch? Also I think you can just call `EVP_DigestVerify` unconditionally here without checking the nids first. That will fallback internally to do `EVP_DigestVerifyUpdate()` and `EVP_DigestVerifyFinal()` if the algorithm doesn't support the one-shot method.